### PR TITLE
move VideoService.cs to Jellyfin.Api

### PR DESCRIPTION
**Pull request by crobibero**
~~Playback doesn't currently work with this PR~~ `/items/{itemId}/PlaybackInfo` wasn't working, so this endpoint may or may not work. I haven't found something that actually uses it.

### CODE REVIEW

File: Jellyfin.Api/Helpers/ProgressiveFileCopier.cs [hunk 1]
```diff
@@ -48,37 +68,123 @@ public ProgressiveFileCopier(IStreamHelper streamHelper, IDirectStreamProvider d
...
+                if (Environment.OSVersion.Platform != PlatformID.Win32NT)
+                {
+                    fileOptions |= FileOptions.Asynchronous;
+                    allowAsyncFileRead = true;
+                }
+
+                await using var inputStream = new FileStream(_path, FileMode.Open, FileAccess.Read, FileShare.ReadWrite, IODefaults.FileStreamBufferSize, fileOptions);
+
+                var eofCount = 0;
+                const int emptyReadLimit = 20;
```
Comment (Bond-009) [inline Jellyfin.Api/Helpers/ProgressiveFileCopier.cs:94]: constants should be PascalCase

File: Jellyfin.Api/Helpers/ProgressiveFileCopier.cs [hunk 2]
```diff
@@ -48,37 +68,123 @@ public ProgressiveFileCopier(IStreamHelper streamHelper, IDirectStreamProvider d
...
             }
+        }
 
-            await using var inputStream = new FileStream(_path, FileMode.Open, FileAccess.Read, FileShare.ReadWrite, 4096, fileOptions);
-            const int emptyReadLimit = 100;
-            var eofCount = 0;
-            while (eofCount < emptyReadLimit)
+        private async Task<int> CopyToInternalAsyncWithSyncRead(Stream source, Stream destination, CancellationToken cancellationToken)
+        {
+            var array = new byte[IODefaults.CopyToBufferSize];
```
Comment (Bond-009) [inline Jellyfin.Api/Helpers/ProgressiveFileCopier.cs:138]: use `ArrayPool`

File: Jellyfin.Api/Helpers/ProgressiveFileCopier.cs [hunk 3]
```diff
@@ -48,37 +68,123 @@ public ProgressiveFileCopier(IStreamHelper streamHelper, IDirectStreamProvider d
...
+                {
+                    await _directStreamProvider.CopyToAsync(outputStream, cancellationToken).ConfigureAwait(false);
+                    return;
+                }
+
+                var fileOptions = FileOptions.SequentialScan;
+                var allowAsyncFileRead = false;
+
+                // use non-async filestream along with read due to https://github.com/dotnet/corefx/issues/6039
+                if (Environment.OSVersion.Platform != PlatformID.Win32NT)
```
Comment (Bond-009) [inline Jellyfin.Api/Helpers/ProgressiveFileCopier.cs:85]: Not sure if `RuntimeInformation.IsOSPlatform` is better here
https://docs.microsoft.com/en-us/dotnet/api/system.environment.osversion?view=netframework-4.7#remarks

File: Jellyfin.Api/Helpers/ProgressiveFileCopier.cs [hunk 4]
```diff
@@ -48,37 +68,123 @@ public ProgressiveFileCopier(IStreamHelper streamHelper, IDirectStreamProvider d
...
                 }
-                else
+            }
+
+            return totalBytesRead;
+        }
+
+        private async Task<int> CopyToInternalAsync(Stream source, Stream destination, CancellationToken cancellationToken)
+        {
+            var array = new byte[IODefaults.CopyToBufferSize];
```
Comment (Bond-009) [inline Jellyfin.Api/Helpers/ProgressiveFileCopier.cs:165]: Use `ArrayPool`

File: Jellyfin.Api/Helpers/ProgressiveFileCopier.cs (hunk 2 above)
Comment (crobibero) [reply Jellyfin.Api/Helpers/ProgressiveFileCopier.cs:138]: How would I use `ArrayPool` properly here?

File: Jellyfin.Api/Helpers/ProgressiveFileCopier.cs (hunk 1 above)
Comment (crobibero) [reply Jellyfin.Api/Helpers/ProgressiveFileCopier.cs:94]: My IDE says local constants should be camelCase

Comment (Bond-009) [reply Jellyfin.Api/Helpers/ProgressiveFileCopier.cs:94]: Your IDE isn't following the `.editorconfig` then.
https://github.com/jellyfin/jellyfin/blob/master/.editorconfig#L77-L85

File: Jellyfin.Api/Helpers/ProgressiveFileCopier.cs (hunk 2 above)
Comment (Bond-009) [reply Jellyfin.Api/Helpers/ProgressiveFileCopier.cs:138]: Example: https://github.com/jellyfin/jellyfin/blob/master/Emby.Server.Implementations/HttpServer/RangeRequestWriter.cs#L186
(Also: code duplication)

File: Jellyfin.Api/Controllers/VideosController.cs [hunk 5]
```diff
@@ -200,5 +261,262 @@ public ActionResult MergeVersions([FromQuery] string? itemIds)
...
+            [FromQuery] string? subtitleCodec,
+            [FromQuery] string? transcodingReasons,
+            [FromQuery] int? audioStreamIndex,
+            [FromQuery] int? videoStreamIndex,
+            [FromQuery] EncodingContext context,
+            [FromQuery] Dictionary<string, string> streamOptions)
+        {
+            var isHeadRequest = Request.Method == System.Net.WebRequestMethods.Http.Head;
+            var cancellationTokenSource = new CancellationTokenSource();
+            var streamingRequest = new StreamingRequestDto
```
Comment (Ullmie02) [inline Jellyfin.Api/Controllers/VideosController.cs:376]:
```suggestion
            var streamingRequest = new VideoRequestDto
```
I think this is the reason why the video gets transcoded to audio only.
This has to be `VideoRequestDto`. Inside of `StreamingHelpers.GetStreamingState` the supported video codecs are only added if `streamingRequest `is a `VideoRequestDto`.

**Comment by Ullmie02** [comment]
Change from `FileStreamResponseHelper.GetStaticFileResult` the current last 2 lines to `return controller.PhysicalFile(path, contentType);` to get video playback working.

Tested with current web master and 720p H264 mkv file

**Pull request merged**